resource-thief.c

```diff
@@ -0,0 +1,361 @@
+	struct wl_display *display;
+	struct wl_registry *registry;
+	struct wl_shm *shm;
+	struct wl_compositor *compositor;
+	struct zwp_linux_dmabuf_v1 *linux_dmabuf;
+	struct wl_shm_pool **pools;
+	struct wl_region **regions;
+	struct wl_buffer **buffers;
+};
+
+void destroy_connection(struct connection *conn) {
+	if (!conn) {
+		return;
+	}
+	if (conn->pools) {
+		for (uint32_t i = 0; i < conn->resource_count; i++) {
+			if (conn->pools[i]) {
+				wl_shm_pool_destroy(conn->pools[i]);
+			}
+		}
+	}
+	if (conn->buffers) {
+		for (uint32_t i = 0; i < conn->resource_count; i++) {
+			if (conn->buffers[i]) {
+				wl_buffer_destroy(conn->buffers[i]);
+			}
+		}
+	}
+	if (conn->regions) {
+		for (uint32_t i = 0; i < (conn->resource_count / POINTS_PER_REGION + 1); i++) {
+			if (conn->regions[i]) {
+				wl_region_destroy(conn->regions[i]);
+			}
+		}
+	}
+	if (conn->shm) {
+		wl_shm_destroy(conn->shm);
+	}
+	if (conn->compositor) {
+		wl_compositor_destroy(conn->compositor);
+	}
+	if (conn->linux_dmabuf) {
+		zwp_linux_dmabuf_v1_destroy(conn->linux_dmabuf);
+	}
+	if (conn->registry) {
+		wl_registry_destroy(conn->registry);
+	}
+	if (conn->display) {
+		wl_display_disconnect(conn->display);
+	}
+	free(conn->pools);
+	free(conn);
+}
+
+
+static void reg_global(void *data, struct wl_registry *registry,
+		uint32_t num, const char *str, uint32_t version) {
+	struct connection *conn = (struct connection *)data;
+	if (!strcmp(str, "wl_shm")) {
+		conn->shm = wl_registry_bind(registry, num,
+			&wl_shm_interface, 1);
+	}
+	if (!strcmp(str, "wl_compositor")) {
+		conn->compositor = wl_registry_bind(registry, num,
+			&wl_compositor_interface, 1);
+	}
+	if (!strcmp(str, "zwp_linux_dmabuf_v1") && version >= 2) {
+		conn->linux_dmabuf = wl_registry_bind(registry, num,
+			&zwp_linux_dmabuf_v1_interface, 2);
+	}
+}
+static void reg_global_remove(void *data, struct wl_registry *wl_registry, uint32_t name) {
+	(void)data;
+	(void)wl_registry;
+	(void)name;
+}
+
+static const struct wl_registry_listener reg_listener = {
+	.global = &reg_global,
+	.global_remove = &reg_global_remove,
+};
+
+struct connection *consume(uint32_t resource_count, enum mode mode, int fd) {
+	const char *resource = "unknown things";
+	if (mode == CONSUME_FDS) {
+		resource = "linux dmabuf fds";
+	} else if (mode == CONSUME_MEMORY) {
+		resource = "wl_region points";
+	} else if (mode == CONSUME_MAPS) {
+		resource = "shm pool mappings";
+	}
+
+	fprintf(stderr, "Trying to create %"PRIu32" %s\n", resource_count, resource);
+
+	struct connection *conn = calloc(1, sizeof(struct connection));
+	if (!conn) {
+			return NULL;
+	}
+	conn->resource_count = resource_count;
+	conn->display = wl_display_connect(NULL);
+	if (!conn->display) {
+		goto fail;
+	}
+	conn->registry = wl_display_get_registry(conn->display);
+	wl_registry_add_listener(conn->registry, &reg_listener, conn);
+	wl_display_roundtrip(conn->display);
+	wl_display_roundtrip(conn->display);
+
+	switch (mode) {
+	case CONSUME_MAPS: {
```

Comment:
I personally prefer `case XXX:;` to avoid weird indentation

Reply:
I have dropped the braces. Since there is no variable declaration immediately following the case label, a trailing semicolon is unnecessary.